tests/mapper/test_link_mappers.py

```diff
@@ -783,3 +783,390 @@ def test_Attri2VecLinkGenerator_unsupervisedSampler_sample_generation(self):
+
+        generator = DirectedGraphSAGELinkGenerator(
+            G,
+            batch_size=self.batch_size,
+            in_samples=self.in_samples,
+            out_samples=self.out_samples,
+        )
+        mapper = generator.flow(G.edges(), edge_labels)
+        assert generator.batch_size == self.batch_size
+        assert mapper.data_size == G.number_of_edges()
+        assert len(mapper.ids) == G.number_of_edges()
+
+    def test_DirectedGraphSAGELinkGenerator_1(self):
+
+        G = example_graph(feature_size=self.n_feat, is_directed=True)
+        data_size = G.number_of_edges()
+        edge_labels = [0] * data_size
+
+        mapper = DirectedGraphSAGELinkGenerator(
+            G,
+            batch_size=self.batch_size,
+            in_samples=self.in_samples,
+            out_samples=self.out_samples,
+        ).flow(G.edges(), edge_labels)
+
+        assert len(mapper) == 2
+
+        for batch in range(len(mapper)):
+            nf, nl = mapper[batch]
+
+            assert len(nf) == 2 ** (len(self.in_samples) + 2) - 2
+
+            for ii in range(len(nf) // 2):
+                assert nf[2 * ii].shape == nf[2 * ii + 1].shape
+
+            # head nodes
+            for ii in range(2):
+                assert nf[ii].shape == (min(self.batch_size, data_size), 1, self.n_feat)
+
+            # first hop in samples
+            for ii in range(2, 4):
+                assert nf[ii].shape == (
+                    min(self.batch_size, data_size),
+                    self.in_samples[0],
+                    self.n_feat,
+                )
+
+            for ii in range(4, 6):
+                assert nf[ii].shape == (
+                    min(self.batch_size, data_size),
+                    self.out_samples[0],
+                    self.n_feat,
+                )
+
+            for ii in range(6, 8):
+                assert nf[ii].shape == (
+                    min(self.batch_size, data_size),
+                    self.in_samples[0] * self.in_samples[1],
+                    self.n_feat,
+                )
+
+            for ii in range(8, 10):
+                assert nf[ii].shape == (
+                    min(self.batch_size, data_size),
+                    self.in_samples[0] * self.out_samples[1],
+                    self.n_feat,
+                )
+
+            for ii in range(10, 12):
+                assert nf[ii].shape == (
+                    min(self.batch_size, data_size),
+                    self.out_samples[0] * self.in_samples[1],
+                    self.n_feat,
+                )
+
+            for ii in range(12, 14):
+                assert nf[ii].shape == (
+                    min(self.batch_size, data_size),
+                    self.out_samples[0] * self.out_samples[1],
+                    self.n_feat,
+                )
+
+        with pytest.raises(IndexError):
+            nf, nl = mapper[2]
+
+    def test_DirectedGraphSAGELinkGenerator_shuffle(self):
```

Comment:
I think the word `DirectedGraphSAGELinkGenerator` can be removed from pretty much all of these test methods - it's already effectively namespaced by the class name
```suggestion
    def test_shuffle(self):
```

Reply:
Instead of the nested function in here, you could use `@pytest.mark.parametrize`
```suggestion
    @pytest.mark.parametrize("shuffle", [True, False])
    def test_shuffle(self, shuffle):
```